New widget: Scan/indexing status

mnpqraven opened the pull request:
while troubleshooting #301 i realized rmpc doesn't have an indicator for db indexing so i think a small loading block would be a nice addition to notify users. I'm still reading the codebase so haven't implemented everything and want to hear your inputs on stuff

todos

- [x] keybinds: planning to have `u` for `rmpc update` and `U` for `rmpc rescan`
- [x] block characters customization: should we allow users to set their own loading characters or is the built-ins enough? (customization in separate PR)
- [x] the loading would pause if the song is also paused rn, make loading state stateful
- [x] the directory, etc. tab also don't get updated while the library is indexing, this list should be updated along with the loading symbols in the event loop i think (not worrying about this for now, i'll probably make a separate PR)
- [x] test + documentation

mierak commented:
> keybinds: planning to have u for rmpc update and U for rmpc rescan

That should be fairly easy to do. I wouldn't bind it in any way to the widget and simply make those actions available as a [GlobalAction]( https://github.com/mierak/rmpc/blob/01a5562b9184233efc6d0bac19d01ae692ab0ba5/src/config/keys/actions.rs#L12). The protocol-level command is is already present in the client so that is not an issue either.

> block characters customization: should we allow users to set their own loading characters or is the built-ins enough

I am fine with either. but allowing config and doing it the same way like its done for labels in [Repeat](https://github.com/mierak/rmpc/blob/01a5562b9184233efc6d0bac19d01ae692ab0ba5/src/config/theme/properties.rs#L47), `Random`, etc would not hurt either.

> currently i just copy-pasted the volume widget and haven't gotten around the widget lifetime 'song (the loading would pause if the song is also pause, want to hear how you want to deal with this)

I dont think you have to tie the widget to the song in any way. Rmpc only renders when something has changed. The render loop while a song is playing is done "artificially" via the scheduler [here](https://github.com/mierak/rmpc/blob/01a5562b9184233efc6d0bac19d01ae692ab0ba5/src/core/event_loop.rs#L252C41-L260C47) which periodically queries MPD which in return triggers a rerender with new values.

You can either schedule a repeated status update job as well or or just use [schedule_replace](https://github.com/mierak/rmpc/blob/01a5562b9184233efc6d0bac19d01ae692ab0ba5/src/core/scheduler/mod.rs#L144) in the `render` function. You would have to render as a stateful widget to get reference to the current status property in the context so you know when to stop the loop as well as to keep the drop guard returned by scheduling a repeated job if you go that route.

mierak commented on src/ui/widgets/scan_status.rs:
```diff
@@ -0,0 +1,105 @@
+            return String::new();
+        }
+
+        // TODO: figure out if we have a mechanism to track ticks instead of
```
We dont, but storing instant of the last "start" of the status update spinner and calculating off of that should be enough no?

mierak commented on src/ui/panes/mod.rs:
```diff
@@ -986,6 +991,15 @@ impl Property<PropertyKind> {
                         },
                     ]))
                 }
+                WidgetProperty::ScanStatus => {
+                    // FIXME: state not persisting, this is wrong and only here
+                    // to make compiler happy
+                    let mut state = ScanStatusState::new(status.updating_db);
+                    Some(Either::Left(Span::styled(
+                        ScanStatus::default().get_str(&mut state),
+                        style,
+                    )))
+                }
```
The UI state is stored in the panes themselves which is not viable here. Instead store/derive this stuff from the context. While typing this I realized this is a bit more involved so let me help, bear with me please.

### 1. Handle `database` idle event
[MPD docs](https://mpd.readthedocs.io/en/latest/protocol.html#querying-mpd-s-status) state: "update: a database update has started or finished. If the database was modified during the update, the database event is also emitted.". We can use this to store the start timestamp and derive whether the update started or finifhed
<details>
<summary>diff</summary>

```diff
diff --git a/src/context.rs b/src/context.rs
index 74acae6..10b510e 100644
--- a/src/context.rs
+++ b/src/context.rs
@@ -1,4 +1,4 @@
-use std::{cell::Cell, collections::HashSet, ops::AddAssign};
+use std::{cell::Cell, collections::HashSet, ops::AddAssign, time::Instant};
 
 use anyhow::{Result, bail};
 use bon::bon;
@@ -31,6 +31,7 @@ pub struct AppContext {
     pub(crate) status: Status,
     pub(crate) queue: Vec<Song>,
     pub(crate) supported_commands: HashSet<String>,
+    pub(crate) db_update_start: Option<Instant>,
     #[debug(skip)]
     pub(crate) app_event_sender: Sender<AppEvent>,
     #[debug(skip)]
@@ -90,6 +91,7 @@ impl AppContext {
             needs_render: Cell::new(false),
             should_fetch_stickers: sticker_support_needed,
             rendered_frames: 0,
+            db_update_start: None,
         })
     }
 
diff --git a/src/core/event_loop.rs b/src/core/event_loop.rs
index 8ab8121..7b1bfec 100644
--- a/src/core/event_loop.rs
+++ b/src/core/event_loop.rs
@@ -28,8 +28,12 @@ use crate::{
         lrc::get_lrc_path,
         macros::{status_error, status_warn},
         mpd_query::{
-            EXTERNAL_COMMAND, GLOBAL_QUEUE_UPDATE, GLOBAL_STATUS_UPDATE, GLOBAL_VOLUME_UPDATE,
-            MpdQueryResult, run_status_update,
+            EXTERNAL_COMMAND,
+            GLOBAL_QUEUE_UPDATE,
+            GLOBAL_STATUS_UPDATE,
+            GLOBAL_VOLUME_UPDATE,
+            MpdQueryResult,
+            run_status_update,
         },
     },
     ui::{KeyHandleResult, Ui, UiAppEvent, UiEvent, modals::info_modal::InfoModal},
@@ -237,9 +241,26 @@ fn main_task<B: Backend + std::io::Write>(
                             let current_song_id =
                                 context.find_current_song_in_queue().map(|(_, song)| song.id);
                             let current_status = context.status.state;
+                            let current_updating_db = context.status.updating_db;
                             context.status = status;
                             let mut song_changed = false;
 
+                            match (current_updating_db, context.status.updating_db) {
+                                (None, Some(_)) => {
+                                    // update of db started
+                                    context.db_update_start = Some(std::time::Instant::now());
+                                }
+                                (Some(_), None) => {
+                                    // update of db ended
+                                    context.db_update_start = None;
+                                }
+                                _ => {
+                                    // no change
+                                }
+                            }
+
                             match context.status.state {
                                 State::Play => {
                                     if current_status != context.status.state {
@@ -476,8 +497,14 @@ fn handle_idle_event(event: IdleEvent, context: &AppContext, result_ui_evs: &mut
                 Ok(MpdQueryResult::Queue(client.playlist_info(fetch_stickers)?))
             });
         }
+        IdleEvent::Database => {
+            context
+                .query()
+                .id(GLOBAL_STATUS_UPDATE)
+                .replace_id("status")
+                .query(move |client| Ok(MpdQueryResult::Status(client.get_status()?)));
+        }
         IdleEvent::StoredPlaylist => {}
-        IdleEvent::Database => {}
         IdleEvent::Update => {}
         IdleEvent::Output
         | IdleEvent::Partition

```
</details>

### 2. update the widget
Change the widget to not store stuff in its state and instead read start timestamp from context. Use it to derive index for the loading chars array. Sorry, this one is leaves some work to do
<details>
<summary>diff</summary>

```diff
diff --git a/src/ui/widgets/scan_status.rs b/src/ui/widgets/scan_status.rs
index 7299f93..217f190 100644
--- a/src/ui/widgets/scan_status.rs
+++ b/src/ui/widgets/scan_status.rs
@@ -1,3 +1,5 @@
+use std::time::Instant;
+
 use ratatui::{
     prelude::Alignment,
     style::Style,
@@ -8,21 +10,22 @@ use super::get_line_offset;
 
 const DEFAULT_LOADING_CHARS: [&str; 8] = ["⣻", "⣽", "⣾", "⣷", "⣯", "⣟", "⡿", "⢿"];
 
+// probably get rid of the widget state altogether
 #[derive(Default, Debug)]
 pub struct ScanStatusState {
-    pub updating: bool,
-    pub symbol_index: usize,
+    // pub updating: bool,
+    // pub symbol_index: usize,
 }
 
-impl ScanStatusState {
-    pub fn cycle_load_symbol(&mut self) {
-        self.symbol_index = (self.symbol_index + 1) % DEFAULT_LOADING_CHARS.len();
-    }
-}
+// impl ScanStatusState {
+//     pub fn cycle_load_symbol(&mut self) {
+//         self.symbol_index = (self.symbol_index + 1) %
+// DEFAULT_LOADING_CHARS.len();     }
+// }
 
 impl ScanStatusState {
-    pub fn new(updating: Option<u32>) -> Self {
-        Self { updating: updating.is_some(), symbol_index: 0 }
+    pub fn new() -> Self {
+        Self {}
     }
 }
 
@@ -31,16 +34,21 @@ pub struct ScanStatus<'a> {
     block: Option<Block<'a>>,
     alignment: Alignment,
     style: Style,
+    update_start: Option<Instant>,
 }
 
-impl Default for ScanStatus<'_> {
-    fn default() -> Self {
-        Self { block: None, alignment: Alignment::Left, style: Style::default() }
-    }
-}
+// impl Default for ScanStatus<'_> {
+//     fn default() -> Self {
+//         Self { block: None, alignment: Alignment::Left, style:
+// Style::default() }     }
+// }
 
 #[allow(dead_code)]
 impl<'a> ScanStatus<'a> {
+    pub fn new(update_start: Option<Instant>) -> Self {
+        Self { block: None, alignment: Alignment::Left, style: Style::default(), update_start }
+    }
+
     pub fn block(mut self, block: Block<'a>) -> Self {
         self.block = Some(block);
         self
@@ -58,13 +66,17 @@ impl<'a> ScanStatus<'a> {
 
     /// get updating symbol, this symbol rotates in set inverval if the db is
     /// scanning
-    pub fn get_str(&mut self, state: &mut ScanStatusState) -> String {
-        if !state.updating {
+    pub fn get_str(&mut self) -> String {
+        let Some(start) = self.update_start else {
             return String::new();
-        }
+        };
         // SAFETY: module of len guarantees the index is always inbound
-        let t = unsafe { DEFAULT_LOADING_CHARS.get_unchecked(state.symbol_index) };
-        state.cycle_load_symbol();
+        // let t = unsafe { DEFAULT_LOADING_CHARS.get_unchecked(state.symbol_index) };
+        // state.cycle_load_symbol();
+
+        let t = DEFAULT_LOADING_CHARS
+            .get(start.elapsed().as_millis() as usize / 250 % DEFAULT_LOADING_CHARS.len())
+            .unwrap_or(&"");
         format!(" {t} ")
     }
 }
@@ -93,6 +105,7 @@ impl StatefulWidget for &mut ScanStatus<'_> {
 
         let left_offset = get_line_offset(3, area.width, self.alignment);
 
-        buf.set_string(area.left() + left_offset, area.top(), self.get_str(state), self.style);
+        log::debug!(str:? = self.get_str(); "str");
+        buf.set_string(area.left() + left_offset, area.top(), self.get_str(), self.style);
     }
 }
```
</details>

The rest is mostly just follow the compiler rodeo, it is a change that multiple parts touch so unfortunately its quite a bit, but nothing too hard. The tests are a bit annyoing tho.

<details>
<summary>diff</summary>

```diff
diff --git a/src/tests/fixtures/mod.rs b/src/tests/fixtures/mod.rs
index 86a0a4f..af2f765 100644
--- a/src/tests/fixtures/mod.rs
+++ b/src/tests/fixtures/mod.rs
@@ -57,6 +57,7 @@ pub fn app_context(
         should_fetch_stickers: false,
         rendered_frames: 0,
         scheduler,
+        db_update_start: None,
     }
 }
 
diff --git a/src/ui/panes/mod.rs b/src/ui/panes/mod.rs
index abe5574..c34e139 100644
--- a/src/ui/panes/mod.rs
+++ b/src/ui/panes/mod.rs
@@ -28,10 +28,7 @@ use unicase::UniCase;
 use self::{frame_count::FrameCountPane, logs::LogsPane};
 use super::{
     UiEvent,
-    widgets::{
-        scan_status::{ScanStatus, ScanStatusState},
-        volume::Volume,
-    },
+    widgets::{scan_status::ScanStatus, volume::Volume},
 };
 use crate::{
     MpdQueryResult,
@@ -41,14 +38,18 @@ use crate::{
         theme::{
             SymbolsConfig,
             properties::{
-                Property, PropertyKind, PropertyKindOrText, SongProperty, StatusProperty,
+                Property,
+                PropertyKind,
+                PropertyKindOrText,
+                SongProperty,
+                StatusProperty,
                 WidgetProperty,
             },
         },
     },
     context::AppContext,
     mpd::{
-        commands::{Song, State, Status, status::OnOffOneshot, volume::Bound},
+        commands::{Song, State, status::OnOffOneshot, volume::Bound},
         mpd_client::Tag,
     },
     shared::{ext::duration::DurationExt, key_event::KeyEvent, mouse_event::MouseEvent},
@@ -533,16 +534,13 @@ pub(crate) mod browser {
 
             input.sort();
 
-            assert_eq!(
-                input,
-                vec![
-                    DirOrSong::Dir { name: "a".to_owned(), full_path: String::new() },
-                    DirOrSong::Dir { name: "z".to_owned(), full_path: String::new() },
-                    DirOrSong::Song(Song::default()),
-                    DirOrSong::Song(Song::default()),
-                    DirOrSong::Song(Song::default()),
-                ]
-            );
+            assert_eq!(input, vec![
+                DirOrSong::Dir { name: "a".to_owned(), full_path: String::new() },
+                DirOrSong::Dir { name: "z".to_owned(), full_path: String::new() },
+                DirOrSong::Song(Song::default()),
+                DirOrSong::Song(Song::default()),
+                DirOrSong::Song(Song::default()),
+            ]);
         }
     }
 }
@@ -842,19 +840,20 @@ impl Property<PropertyKind> {
     fn default_as_span<'song: 's, 's>(
         &'s self,
         song: Option<&'song Song>,
-        status: &'song Status,
+        context: &'song AppContext,
         tag_separator: &str,
     ) -> Option<Either<Span<'s>, Vec<Span<'s>>>> {
-        self.default.as_ref().and_then(|p| p.as_span(song, status, tag_separator))
+        self.default.as_ref().and_then(|p| p.as_span(song, context, tag_separator))
     }
 
     pub fn as_span<'song: 's, 's>(
         &'s self,
         song: Option<&'song Song>,
-        status: &'song Status,
+        context: &'song AppContext,
         tag_separator: &str,
     ) -> Option<Either<Span<'s>, Vec<Span<'s>>>> {
         let style = self.style.unwrap_or_default();
+        let status = &context.status;
         match &self.kind {
             PropertyKindOrText::Text(value) => Some(Either::Left(Span::styled(value, style))),
             PropertyKindOrText::Sticker(key) => {
@@ -863,17 +862,17 @@ impl Property<PropertyKind> {
                 {
                     Some(Either::Left(Span::styled(sticker, style)))
                 } else {
-                    self.default_as_span(song, status, tag_separator)
+                    self.default_as_span(song, context, tag_separator)
                 }
             }
             PropertyKindOrText::Property(PropertyKind::Song(property)) => {
                 if let Some(song) = song {
                     song.format(property, tag_separator).map_or_else(
-                        || self.default_as_span(Some(song), status, tag_separator),
+                        || self.default_as_span(Some(song), context, tag_separator),
                         |s| Some(Either::Left(Span::styled(s, style))),
                     )
                 } else {
-                    self.default_as_span(song, status, tag_separator)
+                    self.default_as_span(song, context, tag_separator)
                 }
             }
             PropertyKindOrText::Property(PropertyKind::Status(s)) => match s {
@@ -959,11 +958,11 @@ impl Property<PropertyKind> {
                     .unwrap_or(style),
                 ))),
                 StatusProperty::Bitrate => status.bitrate.as_ref().map_or_else(
-                    || self.default_as_span(song, status, tag_separator),
+                    || self.default_as_span(song, context, tag_separator),
                     |v| Some(Either::Left(Span::styled(v.to_string(), style))),
                 ),
                 StatusProperty::Crossfade => status.xfade.as_ref().map_or_else(
-                    || self.default_as_span(song, status, tag_separator),
+                    || self.default_as_span(song, context, tag_separator),
                     |v| Some(Either::Left(Span::styled(v.to_string(), style))),
                 ),
             },
@@ -991,20 +990,15 @@ impl Property<PropertyKind> {
                         },
                     ]))
                 }
-                WidgetProperty::ScanStatus => {
-                    // FIXME: state not persisting, this is wrong and only here
-                    // to make compiler happy
-                    let mut state = ScanStatusState::new(status.updating_db);
-                    Some(Either::Left(Span::styled(
-                        ScanStatus::default().get_str(&mut state),
-                        style,
-                    )))
-                }
+                WidgetProperty::ScanStatus => Some(Either::Left(Span::styled(
+                    ScanStatus::new(context.db_update_start).get_str(),
+                    style,
+                ))),
             },
             PropertyKindOrText::Group(group) => {
                 let mut buf = Vec::new();
                 for format in group {
-                    match format.as_span(song, status, tag_separator) {
+                    match format.as_span(song, context, tag_separator) {
                         Some(Either::Left(span)) => buf.push(span),
                         Some(Either::Right(spans)) => buf.extend(spans),
                         None => return None,
@@ -1137,6 +1131,7 @@ mod format_tests {
             style::{Style, Stylize},
             text::Span,
         };
+        use rstest::rstest;
         use test_case::test_case;
 
         use super::*;
@@ -1145,7 +1140,9 @@ mod format_tests {
                 StyleFile,
                 properties::{PropertyKind, StatusProperty, StatusPropertyFile},
             },
+            context::AppContext,
             mpd::commands::{State, Status, Volume, status::OnOffOneshot},
+            tests::fixtures::app_context,
         };
 
         #[test_case(SongProperty::Title, "title")]
@@ -1179,12 +1176,17 @@ mod format_tests {
             assert_eq!(result, Some(expected.to_string()));
         }
 
-        #[test_case(StatusProperty::Volume, "100")]
-        #[test_case(StatusProperty::Elapsed, "2:03")]
-        #[test_case(StatusProperty::Duration, "2:03")]
-        #[test_case(StatusProperty::Crossfade, "3")]
-        #[test_case(StatusProperty::Bitrate, "123")]
-        fn status_property_resolves_correctly(prop: StatusProperty, expected: &str) {
+        #[rstest]
+        #[case(StatusProperty::Volume, "100")]
+        #[case(StatusProperty::Elapsed, "2:03")]
+        #[case(StatusProperty::Duration, "2:03")]
+        #[case(StatusProperty::Crossfade, "3")]
+        #[case(StatusProperty::Bitrate, "123")]
+        fn status_property_resolves_correctly(
+            mut app_context: AppContext,
+            #[case] prop: StatusProperty,
+            #[case] expected: &str,
+        ) {
             let format = Property::<PropertyKind> {
                 kind: PropertyKindOrText::Property(PropertyKind::Status(prop)),
                 style: None,
@@ -1203,7 +1205,7 @@ mod format_tests {
                 ]),
                 stickers: None,
             };
-            let status = Status {
+            app_context.status = Status {
                 volume: Volume::new(123),
                 repeat: true,
                 random: true,
@@ -1217,7 +1219,7 @@ mod format_tests {
                 ..Default::default()
             };
 
-            let result = format.as_span(Some(&song), &status, "");
+            let result = format.as_span(Some(&song), &app_context, "");
 
             assert_eq!(
                 result,
@@ -1225,15 +1227,17 @@ mod format_tests {
             );
         }
 
-        #[test_case("otherplay", "otherstopped", "otherpaused", State::Play, "otherplay")]
-        #[test_case("otherplay", "otherstopped", "otherpaused", State::Pause, "otherpaused")]
-        #[test_case("otherplay", "otherstopped", "otherpaused", State::Stop, "otherstopped")]
+        #[rstest]
+        #[case("otherplay", "otherstopped", "otherpaused", State::Play, "otherplay")]
+        #[case("otherplay", "otherstopped", "otherpaused", State::Pause, "otherpaused")]
+        #[case("otherplay", "otherstopped", "otherpaused", State::Stop, "otherstopped")]
         fn playback_state_label_is_correct(
-            playing_label: &'static str,
-            stopped_label: &'static str,
-            paused_label: &'static str,
-            state: State,
-            expected_label: &str,
+            mut app_context: AppContext,
+            #[case] playing_label: &'static str,
+            #[case] stopped_label: &'static str,
+            #[case] paused_label: &'static str,
+            #[case] state: State,
+            #[case] expected_label: &str,
         ) {
             let format = Property::<PropertyKind> {
                 kind: PropertyKindOrText::Property(PropertyKind::Status(StatusProperty::State {
@@ -1249,9 +1253,9 @@ mod format_tests {
             };
 
             let song = Song { id: 1, file: "file".to_owned(), ..Default::default() };
-            let status = Status { state, ..Default::default() };
+            app_context.status = Status { state, ..Default::default() };
 
-            let result = format.as_span(Some(&song), &status, "");
+            let result = format.as_span(Some(&song), &app_context, "");
 
             assert_eq!(
                 result,
@@ -1259,30 +1263,32 @@ mod format_tests {
             );
         }
 
-        #[test_case(StatusPropertyFile::ConsumeV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, &Status { consume: OnOffOneshot::On, ..Default::default() }, "ye")]
-        #[test_case(StatusPropertyFile::ConsumeV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, &Status { consume: OnOffOneshot::Off, ..Default::default() }, "naw")]
-        #[test_case(StatusPropertyFile::ConsumeV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, &Status { consume: OnOffOneshot::Oneshot, ..Default::default() }, "1111")]
-        #[test_case(StatusPropertyFile::SingleV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, &Status { single: OnOffOneshot::On, ..Default::default() }, "ye")]
-        #[test_case(StatusPropertyFile::SingleV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, &Status { single: OnOffOneshot::Off, ..Default::default() }, "naw")]
-        #[test_case(StatusPropertyFile::SingleV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, &Status { single: OnOffOneshot::Oneshot, ..Default::default() }, "1111")]
-        #[test_case(StatusPropertyFile::RandomV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, &Status { random: true, ..Default::default() }, "ye")]
-        #[test_case(StatusPropertyFile::RandomV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, &Status { random: false, ..Default::default() }, "naw")]
-        #[test_case(StatusPropertyFile::RepeatV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, &Status { repeat: true, ..Default::default() }, "ye")]
-        #[test_case(StatusPropertyFile::RepeatV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, &Status { repeat: false, ..Default::default() }, "naw")]
-        #[test_case(StatusPropertyFile::Consume, &Status { consume: OnOffOneshot::On, ..Default::default() }, "On")]
-        #[test_case(StatusPropertyFile::Consume, &Status { consume: OnOffOneshot::Off, ..Default::default() }, "Off")]
-        #[test_case(StatusPropertyFile::Consume, &Status { consume: OnOffOneshot::Oneshot, ..Default::default() }, "OS")]
-        #[test_case(StatusPropertyFile::Repeat, &Status { repeat: true, ..Default::default() }, "On")]
-        #[test_case(StatusPropertyFile::Repeat, &Status { repeat: false, ..Default::default() }, "Off")]
-        #[test_case(StatusPropertyFile::Random, &Status { random: true, ..Default::default() }, "On")]
-        #[test_case(StatusPropertyFile::Random, &Status { random: false, ..Default::default() }, "Off")]
-        #[test_case(StatusPropertyFile::Single, &Status { single: OnOffOneshot::On, ..Default::default() }, "On")]
-        #[test_case(StatusPropertyFile::Single, &Status { single: OnOffOneshot::Off, ..Default::default() }, "Off")]
-        #[test_case(StatusPropertyFile::Single, &Status { single: OnOffOneshot::Oneshot, ..Default::default() }, "OS")]
+        #[rstest]
+        #[case(StatusPropertyFile::ConsumeV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, Status { consume: OnOffOneshot::On, ..Default::default() }, "ye")]
+        #[case(StatusPropertyFile::ConsumeV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, Status { consume: OnOffOneshot::Off, ..Default::default() }, "naw")]
+        #[case(StatusPropertyFile::ConsumeV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, Status { consume: OnOffOneshot::Oneshot, ..Default::default() }, "1111")]
+        #[case(StatusPropertyFile::SingleV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, Status { single: OnOffOneshot::On, ..Default::default() }, "ye")]
+        #[case(StatusPropertyFile::SingleV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, Status { single: OnOffOneshot::Off, ..Default::default() }, "naw")]
+        #[case(StatusPropertyFile::SingleV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), oneshot_label: "1111".to_string(), on_style: None, off_style: None, oneshot_style: None }, Status { single: OnOffOneshot::Oneshot, ..Default::default() }, "1111")]
+        #[case(StatusPropertyFile::RandomV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, Status { random: true, ..Default::default() }, "ye")]
+        #[case(StatusPropertyFile::RandomV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, Status { random: false, ..Default::default() }, "naw")]
+        #[case(StatusPropertyFile::RepeatV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, Status { repeat: true, ..Default::default() }, "ye")]
+        #[case(StatusPropertyFile::RepeatV2 { on_label: "ye".to_string(), off_label: "naw".to_string(), on_style: None, off_style: None }, Status { repeat: false, ..Default::default() }, "naw")]
+        #[case(StatusPropertyFile::Consume, Status { consume: OnOffOneshot::On, ..Default::default() }, "On")]
+        #[case(StatusPropertyFile::Consume, Status { consume: OnOffOneshot::Off, ..Default::default() }, "Off")]
+        #[case(StatusPropertyFile::Consume, Status { consume: OnOffOneshot::Oneshot, ..Default::default() }, "OS")]
+        #[case(StatusPropertyFile::Repeat, Status { repeat: true, ..Default::default() }, "On")]
+        #[case(StatusPropertyFile::Repeat, Status { repeat: false, ..Default::default() }, "Off")]
+        #[case(StatusPropertyFile::Random, Status { random: true, ..Default::default() }, "On")]
+        #[case(StatusPropertyFile::Random, Status { random: false, ..Default::default() }, "Off")]
+        #[case(StatusPropertyFile::Single, Status { single: OnOffOneshot::On, ..Default::default() }, "On")]
+        #[case(StatusPropertyFile::Single, Status { single: OnOffOneshot::Off, ..Default::default() }, "Off")]
+        #[case(StatusPropertyFile::Single, Status { single: OnOffOneshot::Oneshot, ..Default::default() }, "OS")]
         fn on_off_states_label_is_correct(
-            prop: StatusPropertyFile,
-            status: &Status,
-            expected_label: &str,
+            mut app_context: AppContext,
+            #[case] prop: StatusPropertyFile,
+            #[case] status: Status,
+            #[case] expected_label: &str,
         ) {
             let format = Property::<PropertyKind> {
                 kind: PropertyKindOrText::Property(PropertyKind::Status(prop.try_into().unwrap())),
@@ -1292,23 +1298,26 @@ mod format_tests {
 
             let song = Song { id: 1, file: "file".to_owned(), ..Default::default() };
 
-            let result = format.as_span(Some(&song), status, "");
+            app_context.status = status;
+            let result = format.as_span(Some(&song), &app_context, "");
 
             assert_eq!(result, Some(Either::Left(Span::raw(expected_label))));
         }
 
-        #[test_case(StatusPropertyFile::ConsumeV2 { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), oneshot_style: Some(StyleFile::builder().fg("blue".to_string()).build()), on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, &Status { consume: OnOffOneshot::On, ..Default::default() }, Some(Style::default().red()))]
-        #[test_case(StatusPropertyFile::SingleV2  { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), oneshot_style: Some(StyleFile::builder().fg("blue".to_string()).build()),  on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, &Status { single: OnOffOneshot::On, ..Default::default() }, Some(Style::default().red()))]
-        #[test_case(StatusPropertyFile::RandomV2  { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), on_label: String::new(), off_label: String::new() }, &Status { random: true, ..Default::default() }, Some(Style::default().red()))]
-        #[test_case(StatusPropertyFile::RepeatV2  { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), on_label: String::new(), off_label: String::new() }, &Status { repeat: true, ..Default::default() }, Some(Style::default().red()))]
-        #[test_case(StatusPropertyFile::ConsumeV2 { on_style: None, off_style: None, oneshot_style: None, on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, &Status { consume: OnOffOneshot::On, ..Default::default() }, None)]
-        #[test_case(StatusPropertyFile::SingleV2  { on_style: None, off_style: None, oneshot_style: None, on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, &Status { single: OnOffOneshot::On, ..Default::default() }, None)]
-        #[test_case(StatusPropertyFile::RandomV2  { on_style: None, off_style: None, on_label: String::new(), off_label: String::new() }, &Status { random: true, ..Default::default() }, None)]
-        #[test_case(StatusPropertyFile::RepeatV2  { on_style: None, off_style: None, on_label: String::new(), off_label: String::new() }, &Status { repeat: true, ..Default::default() }, None)]
+        #[rstest]
+        #[case(StatusPropertyFile::ConsumeV2 { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), oneshot_style: Some(StyleFile::builder().fg("blue".to_string()).build()), on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, Status { consume: OnOffOneshot::On, ..Default::default() }, Some(Style::default().red()))]
+        #[case(StatusPropertyFile::SingleV2  { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), oneshot_style: Some(StyleFile::builder().fg("blue".to_string()).build()),  on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, Status { single: OnOffOneshot::On, ..Default::default() }, Some(Style::default().red()))]
+        #[case(StatusPropertyFile::RandomV2  { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), on_label: String::new(), off_label: String::new() }, Status { random: true, ..Default::default() }, Some(Style::default().red()))]
+        #[case(StatusPropertyFile::RepeatV2  { on_style: Some(StyleFile::builder().fg("red".to_string()).build()), off_style: Some(StyleFile::builder().fg("green".to_string()).build()), on_label: String::new(), off_label: String::new() }, Status { repeat: true, ..Default::default() }, Some(Style::default().red()))]
+        #[case(StatusPropertyFile::ConsumeV2 { on_style: None, off_style: None, oneshot_style: None, on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, Status { consume: OnOffOneshot::On, ..Default::default() }, None)]
+        #[case(StatusPropertyFile::SingleV2  { on_style: None, off_style: None, oneshot_style: None, on_label: String::new(), off_label: String::new(), oneshot_label: String::new() }, Status { single: OnOffOneshot::On, ..Default::default() }, None)]
+        #[case(StatusPropertyFile::RandomV2  { on_style: None, off_style: None, on_label: String::new(), off_label: String::new() }, Status { random: true, ..Default::default() }, None)]
+        #[case(StatusPropertyFile::RepeatV2  { on_style: None, off_style: None, on_label: String::new(), off_label: String::new() }, Status { repeat: true, ..Default::default() }, None)]
         fn on_off_oneshot_styles_are_correct(
-            prop: StatusPropertyFile,
-            status: &Status,
-            expected_style: Option<Style>,
+            mut app_context: AppContext,
+            #[case] prop: StatusPropertyFile,
+            #[case] status: Status,
+            #[case] expected_style: Option<Style>,
         ) {
             let format = Property::<PropertyKind> {
                 kind: PropertyKindOrText::Property(PropertyKind::Status(prop.try_into().unwrap())),
@@ -1318,7 +1327,8 @@ mod format_tests {
 
             let song = Song { id: 1, file: "file".to_owned(), ..Default::default() };
 
-            let result = format.as_span(Some(&song), status, "");
+            app_context.status = status;
+            let result = format.as_span(Some(&song), &app_context, "");
 
             dbg!(&result);
             assert_eq!(
diff --git a/src/ui/panes/property.rs b/src/ui/panes/property.rs
index 612d34a..7d57554 100644
--- a/src/ui/panes/property.rs
+++ b/src/ui/panes/property.rs
@@ -37,7 +37,7 @@ impl Pane for PropertyPane<'_> {
         let song = context.find_current_song_in_queue().map(|(_, song)| song);
 
         let line = Line::from(self.content.iter().fold(Vec::new(), |mut acc, val| {
-            match val.as_span(song, &context.status, &context.config.theme.format_tag_separator) {
+            match val.as_span(song, context, &context.config.theme.format_tag_separator) {
                 Some(Either::Left(span)) => acc.push(span),
                 Some(Either::Right(ref mut spans)) => acc.append(spans),
                 None => {}
diff --git a/src/ui/widgets/header.rs b/src/ui/widgets/header.rs
index 8decbec..2d25e53 100644
--- a/src/ui/widgets/header.rs
+++ b/src/ui/widgets/header.rs
@@ -12,7 +12,7 @@ use crate::{
         theme::properties::{Property, PropertyKind},
     },
     context::AppContext,
-    mpd::commands::{Song, Status},
+    mpd::commands::Song,
 };
 
 pub struct Header<'a> {
@@ -41,18 +41,15 @@ impl Widget for Header<'_> {
                 return;
             };
             let template = PropertyTemplates(&config.theme.header.rows[row].left);
-            let widget =
-                template.format(song, &self.context.status, &self.context.config).left_aligned();
+            let widget = template.format(song, self.context, &self.context.config).left_aligned();
             widget.render(left, buf);
 
             let template = PropertyTemplates(&config.theme.header.rows[row].center);
-            let widget =
-                template.format(song, &self.context.status, &self.context.config).centered();
+            let widget = template.format(song, self.context, &self.context.config).centered();
             widget.render(center, buf);
 
             let template = PropertyTemplates(&config.theme.header.rows[row].right);
-            let widget =
-                template.format(song, &self.context.status, &self.context.config).right_aligned();
+            let widget = template.format(song, self.context, &self.context.config).right_aligned();
             widget.render(right, buf);
         }
     }
@@ -60,9 +57,14 @@ impl Widget for Header<'_> {
 
 struct PropertyTemplates<'a>(&'a [Property<PropertyKind>]);
 impl<'a> PropertyTemplates<'a> {
-    fn format(&'a self, song: Option<&'a Song>, status: &'a Status, config: &Config) -> Line<'a> {
+    fn format(
+        &'a self,
+        song: Option<&'a Song>,
+        context: &'a AppContext,
+        config: &Config,
+    ) -> Line<'a> {
         Line::from(self.0.iter().fold(Vec::new(), |mut acc, val| {
-            match val.as_span(song, status, &config.theme.format_tag_separator) {
+            match val.as_span(song, context, &config.theme.format_tag_separator) {
                 Some(Either::Left(span)) => acc.push(span),
                 Some(Either::Right(ref mut spans)) => acc.append(spans),
                 None => {}
```
</details>

You do not have to follow apply the diffs exactly as they are, but its what I would have done here.

mierak commented on src/core/event_loop.rs:
```diff
@@ -80,14 +76,12 @@ fn main_task<B: Backend + std::io::Write>(
-            .map(|interval| context.scheduler.repeated(interval, run_status_update));
-    }
+    // Start the periodic status update for playback and db status
+    _update_loop_guard = context
```
Leave this where it was, there is no need to rerender every second constantly and it now triggers `SongChanged` every rerender.

Instead start a separate timer which sends `AppEvent::RequestRender` when the DB update starts and drop its guard when the update ends.

mnpqraven commented:
@mierak thank you so much for the in-depth snippets! I've updated said diffs and updated the markdown documentation, you can take a look at this PR again now.

mnpqraven commented:
test is failing on `config::tests::example_config_equals_default` though. I've use some prettify tools and see that the keybind block in the `ConfigFile` struct is different because they have different orders. Not sure what caused this
![image](https://github.com/user-attachments/assets/160d5a68-6eb3-496f-8cff-b6a4d6fd17e3)

mierak commented:
rmpc uses nightly for formatting so you have to run`cargo +nightly fmt --all`, other than those few things it looks good to me!

mierak commented:
take a look at the comments I left and we can merge

mnpqraven commented:
> take a look at the comments I left and we can merge

i'm sorry which one exactly ? i'm not seeing any new comments

mierak commented:
the three unresolved code comments i left together with the message about formatting, just above that

mierak commented:
My bad sorry, I did not submit the review. I did not realize it works this way :(

mnpqraven commented:
> My bad sorry, I did not submit the review. I did not realize it works this way :(

all good it happens to everybody lol, anyway i've addressed the comments

mierak commented:
great, thank you!
